src/main/scala/com/vngrs/etl/utils/rdd.scala

```diff
@@ -0,0 +1,116 @@
+package com.vngrs.etl.utils
+
+import com.vngrs.etl.exceptions.NoSchemaException
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.StructType
+
+import scala.util.{Failure, Try, Success}
+
+/**
+  * Enriches [[org.apache.spark.rdd.RDD]] with more functionality
+  */
+object rdd {
+
+  /**
+    * Enriches [[org.apache.spark.rdd.RDD]] to have more functionality
+    *
+    * @param rdd [[org.apache.spark.rdd.RDD]]
+    * @tparam T Type of the RDD
+    */
+  implicit final class RichRDD[T](val rdd: RDD[T]) extends AnyVal {
+
+    /**
+      * Checks whether RDD is non empty or not
+      * @return
+      */
+    def nonEmpty(): Boolean = !rdd.isEmpty()
+  }
+
+  /**
+    * Enriches [[org.apache.spark.rdd.RDD]] of [[org.apache.spark.sql.Row]]s
+    * to have more functionality
+    *
+    * @param rdd [[org.apache.spark.rdd.RDD]]
+    */
+  implicit final class RichRowRDD(val rdd: RDD[Row]) extends AnyVal {
+
+    /**
+      * Fetches Schema
+      *
+      * @return [[com.vngrs.etl.utils.rdd.SchemaResult]]
+      */
+    def schema(): SchemaResult = {
+      Try(rdd.first()) match {
+        case Success(r) => Option(r.schema).map(s => Found(s)).getOrElse(NotFound)
+        case Failure(e) => EmptyRdd
+      }
+    }
+  }
+
+  /**
+    * Schema Result for fetching schema in [org.apache.spark.rdd.RDD]] of [[org.apache.spark.sql.Row]]s
+    */
+  sealed abstract class SchemaResult {
+
+    /**
+      * Folds schema results with given event functions
+      *
+      * @param onFound will be called when a schema is found ([[com.vngrs.etl.utils.rdd.Found]])
+      * @param onNotFound will be called when no schema is found ([[com.vngrs.etl.utils.rdd.NotFound]])
+      * @param onEmptyRdd will be called when `rdd` is empty ([[com.vngrs.etl.utils.rdd.EmptyRdd]])
+      * @tparam A Type of the return
+      * @return result of the event functions
+      */
+    def fold[A](onFound: StructType => A, onNotFound: => A, onEmptyRdd: => A): A = {
+      this match {
+        case Found(s) => onFound(s)
+        case NotFound => onNotFound
+        case EmptyRdd => onEmptyRdd
+      }
+    }
+
+    /**
+      * Gets Schema of the first [[org.apache.spark.sql.Row]] as [[scala.Option]]
+      *
+      * @return Schema of the rows if rdd is non empty and rows have a schema, otherwise `None`
+      */
+    def toOption: Option[StructType] = {
+      this match {
+        case Found(s) => Some(s)
+        case _        => None
+      }
+    }
+
+    /**
+      * Gets Schema of the first [[org.apache.spark.sql.Row]] as [[scala.Option]]
+      * or throws [[com.vngrs.etl.exceptions.NoSchemaException]].
+      *
+      * @param msg Exception Message
+      * @return Schema of the rows if rdd is non empty and rows have a schema
+      * @throws com.vngrs.etl.exceptions.NoSchemaException if no schema is found
+      */
+    @throws(classOf[NoSchemaException])
+    def orThrow(msg: String): StructType = toOption match {
```

Comment:
Yes, of course. What I want to prevent is the misuse of `orThrow`. If you think it is okay, I am okay with that.